Model/Api/OAuthRedirect.php

```diff
@@ -35,37 +46,198 @@ class OAuthRedirect implements OAuthRedirectInterface
         }
 
-        $this->response->setHeader('Content-Type', 'application/json');
-        $this->response->setHttpResponseCode(500);
-        $this->response->setBody(json_encode(['message' => 'Not implemented']));
-        $this->response->sendResponse();
+        if ($code === '' || $scope === '' || $state === '') {
+            throw new WebapiException(__('Bad Request'), 0, WebapiException::HTTP_BAD_REQUEST);
+        }
+
+        $oauthConfiguration = $this->ssoHelper->getOAuthConfiguration();
+        $clientID = $oauthConfiguration['clientID'];
+        $clientSecret = $oauthConfiguration['clientSecret'];
+        $boltPublicKey = $oauthConfiguration['boltPublicKey'];
+
+        $token = $this->ssoHelper->exchangeToken($code, $scope, $clientID, $clientSecret);
+        if ($token === null) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        $payload = $this->ssoHelper->parseAndValidateJWT($token, $clientID, $boltPublicKey);
+        if ($payload === null) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        $websiteId = $this->storeManager->getStore()->getWebsiteId();
+        $storeId = $this->storeManager->getStore()->getId();
+
+        $externalCustomerEntity = null;
+        $customer = null;
+        try {
+            $externalCustomerEntity = $this->externalCustomerEntityRepository->getByExternalID($payload['sub']);
+        } catch (NoSuchEntityException $nsee) {
+        }
+        try {
+            if ($externalCustomerEntity !== null) {
+                $customer = $this->customerRepository->getById($externalCustomerEntity->getCustomerID());
+            } else {
+                $customer = $this->customerRepository->get($payload['email'], $websiteId);
+            }
+        } catch (NoSuchEntityException $nsee) {
+        }
+
+        // External customer entity exists, but the customer it links to doesn't
+        // This should happen very rarely, if at all, which is why we notify bugsnag when this happens
+        if ($externalCustomerEntity !== null && $customer === null) {
+            $this->bugsnag->notifyError(
+                'OAuthRedirect',
+                'external customer entity ' . $payload['sub'] . ' linked to nonexistent customer ' . $externalCustomerEntity->getCustomerID()
+            );
+        }
+
+        // If the customer isn't linked and it exists in M2, but the email is not verified, we throw an exception
+        if ($externalCustomerEntity === null && $customer !== null && !$payload['email_verified']) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        try {
+            $customer = $customer ?: $this->createNewCustomer($websiteId, $storeId, $payload);
+            $this->linkAndLogin($payload['sub'], $customer->getId());
```

Comment:
nit: 'sub' is not readable so maybe it makes sense to create local var at line 177

Reply:
will make change